backend/danswer/tools/tool_selection.py

```diff
@@ -0,0 +1,65 @@
+from typing import Any
+
+from danswer.chat.chat_utils import combine_message_chain
+from danswer.configs.model_configs import GEN_AI_HISTORY_CUTOFF
+from danswer.llm.answering.models import PreviousMessage
+from danswer.llm.interfaces import LLM
+from danswer.llm.utils import message_to_string
+from danswer.prompts.constants import GENERAL_SEP_PAT
+from danswer.tools.tool import Tool
+from danswer.utils.logger import setup_logger
+
+logger = setup_logger()
+
+
+SINGLE_TOOL_SELECTION_PROMPT = f"""
+You are an expert at selecting the most useful tool to run for answering the query.
+You will be given a numbered list of tools and their arguments, a message history, and a query.
+You will select a single tool that will be most useful for answering the query.
+Respond with only the number corresponding to the tool you want to use.
+
+Conversation History:
+{GENERAL_SEP_PAT}
+{{chat_history}}
+{GENERAL_SEP_PAT}
+
+Query:
+{{query}}
+
+Tools:
+{{tool_list}}
+
+Respond with EXACTLY and ONLY the number corresponding to the tool you want to use.
+
+Your selection:
+"""
+
+
+def select_single_tool_for_non_tool_calling_llm(
+    tools_and_args: list[tuple[Tool, dict[str, Any]]],
+    history: list[PreviousMessage],
+    query: str,
+    llm: LLM,
+) -> tuple[Tool, dict[str, Any]] | None:
+    if len(tools_and_args) == 1:
+        return tools_and_args[0]
+
+    tool_list_str = "\n".join(
+        f"""{ind}: {tool.name()} ({args}) - {tool.description()}"""
+        for ind, (tool, args) in enumerate(tools_and_args)
+    )
+
+    history_str = combine_message_chain(
+        messages=history,
+        token_limit=GEN_AI_HISTORY_CUTOFF,
+    )
+    prompt = SINGLE_TOOL_SELECTION_PROMPT.format(
+        tool_list=tool_list_str, chat_history=history_str, query=query
+    )
+    output = message_to_string(llm.invoke(prompt))
+    try:
```

Comment:
I think there should be some sort of regex match to try to extract the int or the tool names as some smaller/worse models may fail to do exactly as instructed. We can even just default to the first tool if it completely fails. We saw with Llama2 that it always tried to elaborate on its answers since it was trained to do this for better answer quality (kind of a forced Chain of Thought)

ie. "use tool run_internet_search which is tool 1 because danswer is not a common term and may be found using an internet search"

so we can try to extract the first number within the range, fall back to checking for any of the tool names, fall back to using the first tool